src/main/java/net/dv8tion/jda/client/entities/impl/ApplicationImpl.java

```diff
@@ -0,0 +1,409 @@
+        private String id;
+        private String name;
+        private String token;
+
+        private BotImpl(final JSONObject object)
+        {
+            this.updateFromJson(object);
+        }
+
+        @Override
+        public boolean equals(final Object obj)
+        {
+            return obj instanceof BotImpl && this.id.equals(((BotImpl) obj).id);
+        }
+
+        @Override
+        public Application getApplication()
+        {
+            return ApplicationImpl.this;
+        }
+
+        @Override
+        public String getAvatarId()
+        {
+            return this.avatarId;
+        }
+
+        @Override
+        public String getAvatarUrl()
+        {
+            return this.avatarId == null ? null
+                    : "https://cdn.discordapp.com/avatars/" + this.id + "/" + this.avatarId + ".jpg";
+        }
+
+        @Override
+        public String getDiscriminator()
+        {
+            return this.discriminator;
+        }
+
+        @Override
+        public String getId()
+        {
+            return this.id;
+        }
+
+        @Override
+        public String getInviteUrl(final Collection<Permission> permissions)
+        {
+            return this.getInviteUrl(null, permissions);
+        }
+
+        @Override
+        public String getInviteUrl(final Permission... permissions)
+        {
+            return this.getInviteUrl(null, permissions);
+        }
+
+        @Override
+        public String getInviteUrl(final String guildId, final Collection<Permission> permissions)
+        {
+            return "https://discordapp.com/oauth2/authorize?client_id=" + this.getId() + "&scope=bot"
+                    + (permissions == null || permissions.isEmpty() ? "" : "&permissions=" + Permission.getRaw(permissions))
+                    + (guildId == null ? "" : "&guild_id=" + guildId);
+        }
+
+        @Override
+        public String getInviteUrl(final String guildId, final Permission... permissions)
+        {
+            return "https://discordapp.com/oauth2/authorize?client_id=" + this.getId() + "&scope=bot"
+                    + (permissions == null || permissions.length == 0 ? ""
+                            : "&permissions=" + Permission.getRaw(permissions))
+                    + (guildId == null ? "" : "&guild_id=" + guildId);
+        }
+
+        public JDA getJDA()
+        {
+            return ApplicationImpl.this.getJDA();
+        }
+
+        @Override
+        public String getName()
+        {
+            return this.name;
+        }
+
+        @Override
+        public String getToken()
+        {
+            return this.token;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return this.id.hashCode();
+        }
+
+        @Override
+        public RestAction<Bot> resetToken()
+        {
+            return new RestAction<Bot>(ApplicationImpl.this.api, Route.Applications.RESET_BOT_TOKEN.compile(this.id),
```

Comment:
Construct the route outside the constructor. / use getJDA() to shorten space required.